### DISCUSSION TITLE
Downgrade support for helmCharts

### DISCUSSION THREAD

**Discussion by jambu** [fluxcd/flux2#2529] [Proposals] [answered]
I have multiple helm charts which I am installing with fluxCD. I need the ability to rollback the system as a whole which is downgrading multiple helm charts (applications support downgrade).

I tested that I can downgrade helmCharts and fluxCD takes care of that. however, the whether it is upgrade or downgrade in version, only the post-upgrade hooks are called. not the post-rollback ones.

There is no way in the hook to determine if this is a downgrade and call the downgrade scripts instead of upgrade scripts. From `.Chart.Version` only the current version can be fetched and not the old version to make the comparison as well.

Support for this will help the system downgrade use cases, requesting a feature where this is possible.

**Comment by kingdonb** [comment, marked as answer]
From Helm's perspective, downgrades are not rollbacks – they are upgrades that happen to change the version in a downward direction. Rollbacks return the release to a previous state from `helm history` and are not explicitly about version numbers going up or down. Anyway, I don't think there is any way for Flux itself to trigger a rollback at present other than through failure remediation.

The only way that I can think of to get the rollback behavior out of Flux now, when the Helm release already succeeded and the upgrade completed: is to suspend the HelmRelease, explicitly call `helm rollback` so that the Helm rollback logic is used, then change the `HelmRelease` spec to match the version rolled back to, and finally `flux resume helmrelease` – un-suspend the helmrelease. (That's obviously not Flux rolling back the release, but if the release didn't fail, then failure remediation strategies will not be invoked.)

Flux will perform rollbacks [when the `remediationStrategy` is configured](https://fluxcd.io/docs/components/helm/helmreleases/#configuring-failure-remediation) for rollbacks, and the release has actually failed.

Flux does not perform rollbacks on its own in any other circumstances that I am aware of today. If you want Flux to invoke the failure remediation and for failed releases to automatically trigger a rollback, then you must ensure they fail when the rollback should be exercised rather than succeeding. You can trigger a failure [by using Helm tests](https://fluxcd.io/docs/components/helm/helmreleases/#configuring-helm-test-actions), if you can predict the expected mode of failure well enough to write a test for it and configure that test to run during the install/upgrade rollout as described.

Maybe there is still room to discuss ways that Flux could be induced to manually trigger a rollback after-the-fact, but I'm not sure how to reconcile that against the circumstances you describe. It would have to be a new addition to the API, and it would be tricky to guarantee that the new spec matches the old release when you wanted to "invoke" the rollback activity in a declarative way.

**Reply by jambu** [reply]
Using helm tests is not feasible for my use case since there are integration tests spanning multiple charts, that declare a deployment as success and some tests might be long running.

I will try the suspend/rollback/resume action and validate that workflow for my use case.
